controllers/metal3machine_controller.go

```diff
@@ -151,6 +145,19 @@ func (r *Metal3MachineReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, re
 		return ctrl.Result{}, errors.Wrapf(err, "failed to create helper for managing the machineMgr")
 	}
 
+	// Return early if the BMMachine or Cluster is paused.
+	if util.IsPaused(cluster, capm3Machine) {
+		machineLog.Info("reconciliation is paused for this object")
+		// set pause annotation on associated bmh (if any)
+		err := machineMgr.SetPauseAnnotation(ctx)
+		if err != nil {
+			machineLog.Info("failed to set pause annotation on associated bmh")
+			return ctrl.Result{}, nil
+		}
+
+		return ctrl.Result{Requeue: true, RequeueAfter: requeueAfter}, nil
+	}
+
```

Comment:
you're missing the logic to remove the Pause annotation if the cluster has been unpaused

Reply:
I was also thinking the same. Do we have a logic for that in case of Metal3Machine?

Reply:
we don't need it for metal3machine. If the cluster is unpaused, the next reconciliation loop will be completed

Reply:
I have added the logic.Please review